### PR TITLE
MAGN-10349 String.Remove anomaly

### PR DESCRIPTION
**Pull request by NightHawk-Michael**
### Purpose
References: [MAGN-10349](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-10349)
- Fixed error String.remove anomaly.
- Updated the test case for string remove

### Declarations

Check these if you believe they are true

- [ ] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.

### Reviewers
@ke-yu

### CODE REVIEW

File: src/Libraries/CoreNodes/String.cs [hunk 1]
```diff
@@ -429,7 +430,11 @@ public static string Remove(string str, int startIndex, int? count = null)
 
             if (_count < 0)
             {
-                startIndex += _count;
+                // The input count is negative means the removal process is required operate from right to left.
+                // However, the removal process in this function always operates rightwards (left to right).
+                // Therefore, a conversion for start index needs to be done in order to change
+                // from leftwards removal (right to left) to rightwards removal (left to right).
+                startIndex = startIndex + _count + 1;
```
Comment (ke-yu) [inline src/Libraries/CoreNodes/String.cs]: Does it still work if `startIndex` and `count` are out of range?

Comment (NightHawk-Michael) [inline src/Libraries/CoreNodes/String.cs]: If the startIndex and Count are out of range, the exception of argumentOutOfRange is thrown, the string.remove node in the UI turn yellow with the meaningful error message for users.

File: src/Libraries/CoreNodes/String.cs [hunk 2]
```diff
@@ -437,6 +437,27 @@ public static string Remove(string str, int startIndex, int? count = null)
...
+            {
+                throw new ArgumentOutOfRangeException("count", "Count is out of range!");
+            }
+
+            if (startIndex >= str.Length || startIndex < 0) 
+            {
+                // startIndex of an array must be within the string length. 
+                // If after the conversion of negative startIndex, startIndex is still
+                // less than zero, ArgumentOutOfRangeException is occured.
+                if (startIndex == str.Length && str.Length == 0)
```
Comment (ke-yu) [inline src/Libraries/CoreNodes/String.cs]: '==' or '<' ?

File: src/Libraries/CoreNodes/String.cs [hunk 3]
```diff
@@ -429,9 +430,34 @@ public static string Remove(string str, int startIndex, int? count = null)
...
+                {
+                    return str.Remove(startIndex, _count);
+                }
+                else
+                {
+                    throw new ArgumentOutOfRangeException("startIndex", "startIndex is out of range!");
```
Comment (ke-yu) [inline src/Libraries/CoreNodes/String.cs]: exception message should be localized. That is, should be added to resource file.

File: src/Libraries/CoreNodes/String.cs [hunk 4]
```diff
@@ -429,9 +430,34 @@ public static string Remove(string str, int startIndex, int? count = null)
...
+            }
+
+            if (startIndex >= str.Length || startIndex < 0) 
+            {
+                // startIndex of an array must be within the string length. 
+                // If after the conversion of negative startIndex, startIndex is still
+                // less than zero, ArgumentOutOfRangeException is occured.
+                if (startIndex == str.Length && str.Length == 0)
+                {
+                    return str.Remove(startIndex, _count);
```
Comment (ke-yu) [inline src/Libraries/CoreNodes/String.cs]: Why we need to call `str.Remove()` if `str.Length` is 0?

Comment (NightHawk-Michael) [inline src/Libraries/CoreNodes/String.cs]: @ke-yu In the original test case, there is an edge case test: Assert.AreEqual("", String.Remove("", 0));
So I was thinking of just preserving that so we still allow str.Length is 0.

File: src/Libraries/CoreNodes/String.cs [hunk 5]
```diff
@@ -429,9 +430,35 @@ public static string Remove(string str, int startIndex, int? count = null)
...
+                startIndex = startIndex + _count + 1; 
                 _count *= -1;
             }
+            
+            if(_count > str.Length)
+            {
+                throw new ArgumentOutOfRangeException("count", Properties.Resources.StringRemoveCountOutOfRangeMessage);
+            }
+
+            if (startIndex >= str.Length || startIndex < 0)
```
Comment (ke-yu) [inline src/Libraries/CoreNodes/String.cs]: The logic is a little bit weird here. I'd suggest to split it into two:
```
if (str.Legnth == 0 && startIndex == 0)
{
    return string.Empty;
}

if (startIndex >= str.Length || starIndex < 0)
{
    throw ...
}

...
```

**Pull request merged**